### ISSUE TITLE
'Illegal instruction (core dumped)' when running the examples

### ISSUE THREAD

**Issue by Magalame** [mikeizbicki/simd#4]
When running the examples, some work fine, but it seems like there is an upper bound on the N of the `XN` simd vectors.
For example VU/Float will crash on X16. And VU/Double will crash on X8

Since the Doubles takes twice as much memory as the Floats, there is a clear pattern

**Comment by mikeizbicki** [comment]
Hmm... my best guess is that this is due to your CPU not supporting the
necessary instructions.  IIRC, there's no checks in the code to actually
ensure that the CPU supports the needed instructions.

**Comment by Magalame** [comment]
It seems like you are right. Do you know of any way I could check which instructions are available? I know you're not working on your Haskell projects anymore but just a hint would be amazing !

**Comment by Magalame** [comment]
While I have you here, I have another question if you don't mind. Did you have any specific idea regarding how to tackle vectors whose length isn't a multiple of 2, 4, etc?

**Comment by Drezil** [comment]
> It seems like you are right. Do you know of any way I could check which instructions are available?

Under linux you can just query /proc/cpuinfo (i.e. on my system):
```
cat /proc/cpuinfo | grep flags
flags           : fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx rdtscp lm constant_tsc arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic popcnt tsc_deadline_timer aes xsave avx lahf_lm epb pti ssbd ibrs ibpb stibp tpr_shadow vnmi flexpriority ept vpid xsaveopt dtherm ida arat pln pts md_clear flush_l1d
```

for compiling i think you have to fall back to using flags (like -fsse2 -fssse3 -favx) and then activate/deactivate the instructions in haskell with cpp-pragmas (#ifdef sse2 ... #else .....)

> While I have you here, I have another question if you don't mind. Did you have any specific idea regarding how to tackle vectors whose length isn't a multiple of 2, 4, etc?

https://www.youtube.com/watch?v=KzqNQMpRbac&ab_channel=HSRHochschulef%C3%BCrTechnikRapperswil i found that interesting. Edward Kmett talking about his C++-Adventures with exactly this problem.
Basic gist: chunk your vector to the max amount you have ([i.e. like this](https://hackage.haskell.org/package/split-0.2.3.3/docs/Data-List-Split.html#v:chunksOf)) and for the leftovers either pad them with 0, use a ['do not actually do this on this input'-mask](https://software.intel.com/sites/landingpage/IntrinsicsGuide/#cats=Move&expand=3789) in ASM or do not use SIMD etc. for the leftovers.

**Comment by Magalame** [comment]
Hey @Drezil! 

> Under linux you can just query /proc/cpuinfo (i.e. on my system):

I was wondering if it was possible to do that from the Haskell package itself. So that the end user doesn't have to check anything themselves. Ideally at compile-time we would detect which instructions are available, and then have the #ifdef handle the rest

> Edward Kmett talking about his C++-Adventures with exactly this problem.

Sounds perfect, I'll dig into this as soon as I can!

**Comment by Magalame** [comment]
@Drezil I found a simple, and probably performant solution to the padding problem, although a tad nasty.
`simd` currently uses `unsafeCoerce` when vectorizing `vector`s. 

The idea is that a `vector` can be coerced to `Vector !Int !Int !ByteArray` (http://hackage.haskell.org/package/vector-0.12.0.2/docs/Data-Vector-Primitive.html). The first int being the offset, and the second the length (warning, it seems like there is a mix up in `simd`, since the first int is pattern-matched as `len`, and the second as `off`), then we can just have (for x4 vectors):

```haskell
UnsafePrimVector off len arr = unsafeCoerce initialVector -- force pattern match to extract properties
vectorizedVector = unsafeCoerce $ UnsafePrimVector (off `div`4) (len `div` 4) arr -- reorganize data layout and coerce to appropriate type
```
this is the gist of `unsafeVectorizeUnboxedX4` for example

This obviously works well when `len` and `off` are divisible by 4, but otherwise it is not obvious what might happen. Turns out after a few trials (figuring out the `off`/`len` mix up took me way too much time), that when `len`%4 != 0, we can simply have (let's call this `vectorizeWithPad` for later):

```haskell
UnsafePrimVector off len arr = unsafeCoerce initialVector
vectorizedVector = unsafeCoerce $ UnsafePrimVector (off `div`4) (len `div` 4 + 1) arr
```
and ghc will spontaneously zero-pad the last element. 

Using a very concrete example:
```haskell
v = U.fromList [1,2,3,4,5,6] -- len: 6, off: 0
unsafeVectorizeUnboxedX4 v -- gives [(1,2,3,4)]
vectorizeWithPad v  -- gives [(1,2,3,4),(5,6,0,0)]
```

what I like about this approach:
- simple to understand and implement
- probably performant, I think it doesn't copy the data during the conversion. To be checked though

what I do not like:
- is defaulting to 0 an undefined behaviour and we are just lucky?If it is defined then where?
- we might need an efficient way to keep a track of where does the padding start, when we'll need to unvectorize the vector

todo: how to handle the offset. I think it could be doable to take a slice of `arr`and set the offset to 0

**Comment by Magalame** [comment]
So, to properly deal with the offset, I could see something like:
```haskell
marr = thaw arr
moveByteArray marr 0 marr offset len*(I# (sizeOf# an element of arr))
arrnew = freeze marr
-- should there be a call to resizeMutableByteArray here?
```
then wejust have to go through the initial process, when the length is the only potential problem.

I'm worried about performance though, I think `moveByteArray` might copy the whole array, instead of just moving a pointer. But if it is so, then there is no reason for `copyMutableByteArray`

**Comment by mikeizbicki** [comment]
I'm just curious, are you working on some type of new vector library?

**Comment by Magalame** [comment]
Not quite! This would be way off my abilities, and it seems like `massiv` is already going more or less in that direction with promising results. But I'm trying to improve the performances of linear algebra tools in the Haskell ecosystem. I've been focusing on `dense-linear-algebra`, since it has a very simple back end, which makes it easier to reason about performances. Ideally we'd eventually reach the same or better performances than what `hmatrix` yield. If I can interest you in some benchmarks: https://github.com/Magalame/fastest-matrices

**Comment by mikeizbicki** [comment]
Interesting!  I once went down this path before...  Here's some
unsolicited advice :)

The reason hmatrix is so much faster than the other libraries is that
it's using BLAS under the hood.  I doubt you'll be able to compete
with that for matrices any bigger than say 5x5.  Also, for your
benchmarks, I typically see matrices on the order of 10^3-10^6, so I'd
be more interested in performance at that scale than at the scale you
benchmarked.

Finally, some things that I think Haskell really needs but doesn't
have are better support for sparse matrices and higher order tensors.
There are BLAS equivalents for both of those, and IIRC I once had some
sort of bindings set up for them, but I couldn't come up with an
interface design I was happy with and gave up on the project.

Good luck!  I'd love better Haskell linear alg tools :)

**Comment by Magalame** [comment]
Thanks for the advice, it's always welcome! :) 

Indeed, BLAS is the reason behind `hmatrix`'s speed, but I'm pretty sure we can get very similar performances with pure Haskell. From my tests SIMD alone is a huge step up!

I limited myself to 100*100 because otherwise some libraries would crash. But I guess it'd make more sense to just skip them where they fail

@ocramz is working on some sparse matrix library :)

**Comment by mikeizbicki** [comment]
One thing about the pure Haskell approach that I really like is that then
you could seamlessly use `Rational`s or other numeric types not supported
by BLAS.

**Comment by Magalame** [comment]
Pure haskell approaches also seem to attract more contributors. For example `hmatrix` is unfortunately barely maintained despite being largely used :/
